src/Features/CSharp/Portable/UseExpressionBodyForLambda/UseExpressionBodyForLambdaCodeStyleProvider.cs

```diff
@@ -0,0 +1,209 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.CSharp.CodeStyle;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.UseExpressionBodyForLambda
+{
+    internal partial class UseExpressionBodyForLambdaCodeStyleProvider
+        : AbstractCodeStyleProvider<ExpressionBodyPreference, SyntaxKind, UseExpressionBodyForLambdaCodeStyleProvider>
+    {
+        private static readonly LocalizableString UseExpressionBodyTitle = new LocalizableResourceString(nameof(FeaturesResources.Use_expression_body_for_lambda_expressions), FeaturesResources.ResourceManager, typeof(FeaturesResources));
+        private static readonly LocalizableString UseBlockBodyTitle = new LocalizableResourceString(nameof(FeaturesResources.Use_block_body_for_lambda_expressions), FeaturesResources.ResourceManager, typeof(FeaturesResources));
+
+        public UseExpressionBodyForLambdaCodeStyleProvider() 
+            : base(CSharpCodeStyleOptions.PreferExpressionBodiedLambdaExpressions,
+                   IDEDiagnosticIds.UseExpressionBodyForLambdaExpressionsDiagnosticId,
+                   UseExpressionBodyTitle,
+                   UseExpressionBodyTitle)
+        {
+        }
+
+        private static ExpressionSyntax GetBodyAsExpression(LambdaExpressionSyntax declaration)
+            => declaration.Body as ExpressionSyntax;
+
+        private static bool CanOfferUseExpressionBody(
+            ExpressionBodyPreference preference, LambdaExpressionSyntax declaration)
+        {
+            var userPrefersExpressionBodies = preference != ExpressionBodyPreference.Never;
+            if (!userPrefersExpressionBodies)
```

Comment:
Note the simplicity of this method compared to what it has to be over here: https://github.com/dotnet/roslyn/pull/28765/files#diff-79b0bc1373e80fce4c6740e4326dd50bR30

There's no need for complex and error prone code like:

```
+            // If the user likes expression bodies, then we offer expression bodies from the diagnostic analyzer.
+            // If the user does not like expression bodies then we offer expression bodies from the refactoring provider.
+            // If the analyzer is disabled completely, the refactoring is enabled in both directions.
+            if (userPrefersExpressionBodies == forAnalyzer || (!forAnalyzer && analyzerDisabled))
```

We only get into this codepath in the first place if the user actually has this option enabled in a way such that analyzer should run.